File: binary_search/4_median_of_two_sorted_arrays.py
# https://cloud.tencent.com/developer/article/1594692
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        len1 = len(nums1)
        len2 = len(nums2)
        if len1 > len2:
            return self.findMedianSortedArrays(nums2, nums1)

        left = 0
        right = len1
        while left <= right:
            partition1 = left + (right - left)//2
            partition2 = (len1 + len2)//2 - partition1
            left1 = nums1[partition1-1] if partition1 > 0 else float('-inf')
            right1 = nums1[partition1] if partition1 < len1 else float('inf')
            left2 = nums2[partition2-1] if partition2 > 0 else float('-inf')
            right2 = nums2[partition2] if partition2 < len2 else float('inf')
            if left1 <= right2 and left2 <= right1:
                if (len1 + len2) % 2:
                    return min(right1, right2)
                else:
                    return (max(left1, left2) + min(right1, right2))/2
            elif left1 > right2:
                right = partition1 - 1
            else:
                left = partition1 + 1

        return 0

File: interval_matrix_binaryindextree/4_median_of_two_sorted_arrays.java
// Explanation: https://www.youtube.com/watch?v=LPFhl65R7ww
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int len1 = nums1.length;
        int len2 = nums2.length;

        if (len1 > len2) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int l = 0;
        int r = len1;

        while (l <= r) {
            int partition1 = (l + r) / 2;
            int partition2 = (len1 + len2 + 1) / 2 - partition1;

            int left_max1 = (partition1 == 0) ? Integer.MIN_VALUE : nums1[partition1 - 1];
            int right_min1 = (partition1 == len1) ? Integer.MAX_VALUE : nums1[partition1];

            int left_max2 = (partition2 == 0) ? Integer.MIN_VALUE : nums2[partition2 - 1];
            int right_min2 = (partition2 == len2) ? Integer.MAX_VALUE : nums2[partition2];

            if (left_max1 <= right_min2 && left_max2 <= right_min1) {
                if ((len1 + len2) % 2 == 0) {
                    return ((double) Math.max(left_max1, left_max2) + Math.min(right_min1, right_min2)) / 2;
                }
                else {
                    return ((double) Math.max(left_max1, left_max2));
                }
            }
            else if (left_max1 > right_min2) {
                r = partition1 - 1;
            }
            else {
                l = partition1 + 1;
            }
            }
        return -1;
}}